Goal of this notebook:

Investigate the impact of replacing NA values in the dataset with a placeholder on the classification accuracy with decision tree and random forest.  (XGBoost Random Forest works with NA values natively)

In [1]:
import numpy as np
import pandas as pd
from IPython.core.display import display

pd.set_option('display.max_columns', None)
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import warnings
import classifier_utils
warnings.filterwarnings("ignore", category=UserWarning)

In [2]:
non_features_columns = ["chunk_id", "line_start", "line_end", "line_separator", "kind_conflict", "url", "project"]
non_features_columns.extend(["project_user", "project_name", "path", "file_name", "sha", "leftsha", "rightsha", "basesha"])

In [3]:
selected_dataset = pd.read_csv("../../data/SELECTED_LABELLED_DATASET.csv")
projects = list(selected_dataset['project'].unique())

# Decision Tree

In [4]:
dt = DecisionTreeClassifier(random_state=99)

In [5]:
result_dt = classifier_utils.ProjectsResults(dt, projects, non_features_columns)

In [6]:
report_dt = result_dt.get_report_df(include_overall=True)
report_dt

project  observations  \
0                         CCI-MIT__XCoLab          5512   
1                apache__directory-server           845   
2                        jgralab__jgralab          2072   
3                        Unidata__thredds          1154   
4                        apache__accumulo          4113   
5   CloudStack-extras__CloudStack-archive          1424   
6                         getrailo__railo           815   
7                  Ramblurr__Anki-Android           892   
8                  TeamDev-Ltd__OpenFaces          2979   
9                     apache__lucene-solr          1256   
10                              zkoss__zk          1087   
11                          freenet__fred          1268   
12               hibernate__hibernate-orm          1000   
13                           alexo__wro4j          1663   
14      android__platform_frameworks_base          3557   
15                 eucalyptus__eucalyptus          1020   
16                        atlasapi__atlas          1107   
17                  alkacon__opencms-core           923   
18                   SINTEF-9012__ThingML          1022   
19                  sebastianbenz__Jnario          8298   
20                 elastic__elasticsearch          1246   
21                 eclipse__jetty.project          1537   
22            revolsys__com.revolsys.open           991   
23                                Overall         45781   

    observations (wt NaN)  precision   recall  f1-score  accuracy  \
0                    3757    0.97200  0.97200    0.9720   0.97200   
1                     652    0.92600  0.92800    0.9270   0.92800   
2                    1802    0.83900  0.83900    0.8390   0.83900   
3                     950    0.90700  0.90900    0.9080   0.90900   
4                    3148    0.83800  0.83800    0.8380   0.83800   
5                    1106    0.74800  0.75000    0.7490   0.75000   
6                     572    0.65600  0.65000    0.6520   0.65000   
7                     759    0.67200  0.68200    0.6770   0.68200   
8                    2859    0.96300  0.96400    0.9640   0.96400   
9                     974    0.55800  0.55500    0.5560   0.55500   
10                    881    0.76100  0.75400    0.7570   0.75400   
11                   1012    0.60300  0.60100    0.6020   0.60100   
12                    716    0.51500  0.50800    0.5110   0.50800   
13                   1368    0.49900  0.49900    0.4980   0.49900   
14                   2460    0.77600  0.77600    0.7760   0.77600   
15                    747    0.56000  0.55700    0.5580   0.55700   
16                    782    0.57000  0.57300    0.5710   0.57300   
17                    840    0.94700  0.94900    0.9480   0.94900   
18                    977    0.93700  0.93800    0.9370   0.93800   
19                   8173    0.99800  0.99900    0.9980   0.99900   
20                      0        NaN      NaN       NaN       NaN   
21                      0        NaN      NaN       NaN       NaN   
22                      1        NaN      NaN       NaN       NaN   
23                  34536    0.76225  0.76205    0.7619   0.76205   

    baseline (majority)  improvement  
0               0.57300      0.93300  
1               0.51200      0.85200  
2               0.49100      0.68300  
3               0.77700      0.59400  
4               0.63500      0.55700  
5               0.43700      0.55500  
6               0.37800      0.43800  
7               0.43900      0.43400  
8               0.93800      0.41500  
9               0.26600      0.39400  
10              0.60200      0.38200  
11              0.40400      0.33000  
12              0.31100      0.28600  
13              0.30700      0.27600  
14              0.70800      0.23400  
15              0.42300      0.23200  
16              0.44500      0.23000  
17              0.94400      0.08500  
18              0.93800      0.00000  
19              0.99900     -0.00100  
20           

In [7]:
import importlib
importlib.reload(classifier_utils)
result_dt2 = classifier_utils.ProjectsResults(dt, projects, non_features_columns, replace_na=True)

In [8]:
report_dt2 = result_dt2.get_report_df(include_overall=True)
report_dt2

project  observations  \
0                         CCI-MIT__XCoLab          5512   
1                apache__directory-server           845   
2                        jgralab__jgralab          2072   
3   CloudStack-extras__CloudStack-archive          1424   
4                        Unidata__thredds          1154   
5                        apache__accumulo          4113   
6                         getrailo__railo           815   
7                  Ramblurr__Anki-Android           892   
8                  TeamDev-Ltd__OpenFaces          2979   
9                     apache__lucene-solr          1256   
10                          freenet__fred          1268   
11                 elastic__elasticsearch          1246   
12                           alexo__wro4j          1663   
13                              zkoss__zk          1087   
14               hibernate__hibernate-orm          1000   
15                        atlasapi__atlas          1107   
16                 eclipse__jetty.project          1537   
17      android__platform_frameworks_base          3557   
18                 eucalyptus__eucalyptus          1020   
19            revolsys__com.revolsys.open           991   
20                  alkacon__opencms-core           923   
21                  sebastianbenz__Jnario          8298   
22                   SINTEF-9012__ThingML          1022   
23                                Overall         45781   

    observations (wt NaN)  precision    recall  f1-score  accuracy  \
0                    5512   0.971000  0.971000  0.971000  0.971000   
1                     845   0.902000  0.902000  0.902000  0.902000   
2                    2072   0.823000  0.823000  0.823000  0.823000   
3                    1424   0.778000  0.780000  0.779000  0.780000   
4                    1154   0.903000  0.904000  0.903000  0.904000   
5                    4113   0.852000  0.854000  0.853000  0.854000   
6                     815   0.725000  0.721000  0.723000  0.721000   
7                     892   0.684000  0.674000  0.679000  0.674000   
8                    2979   0.952000  0.952000  0.952000  0.952000   
9                    1256   0.537000  0.538000  0.538000  0.538000   
10                   1268   0.621000  0.621000  0.621000  0.621000   
11                   1246   0.536000  0.534000  0.534000  0.534000   
12                   1663   0.510000  0.507000  0.508000  0.507000   
13                   1087   0.733000  0.730000  0.730000  0.730000   
14                   1000   0.506000  0.506000  0.506000  0.506000   
15                   1107   0.586000  0.584000  0.584000  0.584000   
16                   1537   0.569000  0.568000  0.568000  0.568000   
17                   3557   0.795000  0.793000  0.794000  0.793000   
18                   1020   0.563000  0.558000  0.560000  0.558000   
19                    991   0.755000  0.792000  0.738000  0.792000   
20                    923   0.929000  0.927000  0.928000  0.927000   
21                   8298   0.998000  0.998000  0.998000  0.998000   
22                   1022   0.921000  0.920000  0.921000  0.920000   
23                  45781   0.745609  0.745957  0.744043  0.745957   

    baseline (majority)  improvement  
0              0.454000     0.946000  
1              0.515000     0.798000  
2              0.480000     0.660000  
3              0.453000     0.598000  
4              0.775000     0.573000  
5              0.687000     0.535000  
6              0.459000     0.485000  
7              0.436000     0.421000  
8              0.924000     0.369000  
9              0.289000     0.351000  
10             0.424000     0.342000  
11             0.319000     0.316000  
12             0.310000     0.286000  
13             0.622000     0.285000  
14             0.312000     0.282000  
15             0.455000     0.235000  
16             0.446000     0.221000  
17             0.743000     0.196000  
18             0.463000     0.177000  
19             0.781000    

In [9]:
df_inner_dt = pd.merge(report_dt, report_dt2, on='project', how='inner', suffixes=('_withoutNA', '_withNA'))

accuracy_dt = df_inner_dt.filter(regex=("project|accuracy.*")).copy()
accuracy_dt['improvement'] = accuracy_dt.apply(lambda x: classifier_utils.get_normalized_improvement(x['accuracy_withNA'], x['accuracy_withoutNA']), axis=1)
accuracy_dt

project  accuracy_withoutNA  \
0                         CCI-MIT__XCoLab             0.97200   
1                apache__directory-server             0.92800   
2                        jgralab__jgralab             0.83900   
3                        Unidata__thredds             0.90900   
4                        apache__accumulo             0.83800   
5   CloudStack-extras__CloudStack-archive             0.75000   
6                         getrailo__railo             0.65000   
7                  Ramblurr__Anki-Android             0.68200   
8                  TeamDev-Ltd__OpenFaces             0.96400   
9                     apache__lucene-solr             0.55500   
10                              zkoss__zk             0.75400   
11                          freenet__fred             0.60100   
12               hibernate__hibernate-orm             0.50800   
13                           alexo__wro4j             0.49900   
14      android__platform_frameworks_base             0.77600   
15                 eucalyptus__eucalyptus             0.55700   
16                        atlasapi__atlas             0.57300   
17                  alkacon__opencms-core             0.94900   
18                   SINTEF-9012__ThingML             0.93800   
19                  sebastianbenz__Jnario             0.99900   
20                 elastic__elasticsearch                 NaN   
21                 eclipse__jetty.project                 NaN   
22            revolsys__com.revolsys.open                 NaN   
23                                Overall             0.76205   

    accuracy_withNA  improvement  
0          0.971000    -0.001029  
1          0.902000    -0.028017  
2          0.823000    -0.019070  
3          0.904000    -0.005501  
4          0.854000     0.098765  
5          0.780000     0.120000  
6          0.721000     0.202857  
7          0.674000    -0.011730  
8          0.952000    -0.012448  
9          0.538000    -0.030631  
10         0.730000    -0.031830  
11         0.621000     0.050125  
12         0.506000    -0.003937  
13         0.507000     0.015968  
14         0.793000     0.075893  
15         0.558000     0.002257  
16         0.584000     0.025761  
17         0.927000    -0.023182  
18         0.920000    -0.019190  
19         0.998000    -0.001001  
20         0.534000          NaN  
21         0.568000          NaN  
22         0.792000          NaN  
23         0.745957    -0.021119

In [10]:
accuracy_dt

project  accuracy_withoutNA  \
0                         CCI-MIT__XCoLab             0.97200   
1                apache__directory-server             0.92800   
2                        jgralab__jgralab             0.83900   
3                        Unidata__thredds             0.90900   
4                        apache__accumulo             0.83800   
5   CloudStack-extras__CloudStack-archive             0.75000   
6                         getrailo__railo             0.65000   
7                  Ramblurr__Anki-Android             0.68200   
8                  TeamDev-Ltd__OpenFaces             0.96400   
9                     apache__lucene-solr             0.55500   
10                              zkoss__zk             0.75400   
11                          freenet__fred             0.60100   
12               hibernate__hibernate-orm             0.50800   
13                           alexo__wro4j             0.49900   
14      android__platform_frameworks_base             0.77600   
15                 eucalyptus__eucalyptus             0.55700   
16                        atlasapi__atlas             0.57300   
17                  alkacon__opencms-core             0.94900   
18                   SINTEF-9012__ThingML             0.93800   
19                  sebastianbenz__Jnario             0.99900   
20                 elastic__elasticsearch                 NaN   
21                 eclipse__jetty.project                 NaN   
22            revolsys__com.revolsys.open                 NaN   
23                                Overall             0.76205   

    accuracy_withNA  improvement  
0          0.971000    -0.001029  
1          0.902000    -0.028017  
2          0.823000    -0.019070  
3          0.904000    -0.005501  
4          0.854000     0.098765  
5          0.780000     0.120000  
6          0.721000     0.202857  
7          0.674000    -0.011730  
8          0.952000    -0.012448  
9          0.538000    -0.030631  
10         0.730000    -0.031830  
11         0.621000     0.050125  
12         0.506000    -0.003937  
13         0.507000     0.015968  
14         0.793000     0.075893  
15         0.558000     0.002257  
16         0.584000     0.025761  
17         0.927000    -0.023182  
18         0.920000    -0.019190  
19         0.998000    -0.001001  
20         0.534000          NaN  
21         0.568000          NaN  
22         0.792000          NaN  
23         0.745957    -0.021119

In [19]:
increased_n = len(accuracy_dt[(accuracy_dt['improvement'] > 0) &  (accuracy_dt['project'] != 'Overall')])
decreased_n = len(accuracy_dt[(accuracy_dt['improvement'] < 0) &  (accuracy_dt['project'] != 'Overall')])
print(f'Increased in {increased_n} and decreased in {decreased_n} projects.')

Increased in 8 and decreased in 12 projects.


The overall average accuracy with decision tree decreased from 0.76205 to 0.745957. It decreased in 12 of 20 projects and increased in 8.
On the other hand, inputing a constant to replace the NA values allowed to go from 20 to 23 projects.

# Random forest

In [11]:
model_3 = RandomForestClassifier(random_state=99, n_jobs=5, n_estimators=400, max_features=0.3, min_samples_leaf=1)# Decision Tree

In [12]:
result_dt3 = classifier_utils.ProjectsResults(model_3, projects, non_features_columns, replace_na=False)
report_dt3 = result_dt3.get_report_df(include_overall=True)

In [13]:
report_dt3

project  observations  \
0                         CCI-MIT__XCoLab          5512   
1                apache__directory-server           845   
2                        jgralab__jgralab          2072   
3   CloudStack-extras__CloudStack-archive          1424   
4                        apache__accumulo          4113   
5                        Unidata__thredds          1154   
6                  Ramblurr__Anki-Android           892   
7                         getrailo__railo           815   
8                     apache__lucene-solr          1256   
9                  TeamDev-Ltd__OpenFaces          2979   
10                          freenet__fred          1268   
11                              zkoss__zk          1087   
12               hibernate__hibernate-orm          1000   
13                        atlasapi__atlas          1107   
14                           alexo__wro4j          1663   
15      android__platform_frameworks_base          3557   
16                 eucalyptus__eucalyptus          1020   
17                  alkacon__opencms-core           923   
18                   SINTEF-9012__ThingML          1022   
19                  sebastianbenz__Jnario          8298   
20                 elastic__elasticsearch          1246   
21                 eclipse__jetty.project          1537   
22            revolsys__com.revolsys.open           991   
23                                Overall         45781   

    observations (wt NaN)  precision   recall  f1-score  accuracy  \
0                    3757     0.9720  0.97600    0.9730   0.97600   
1                     652     0.9340  0.93700    0.9360   0.93700   
2                    1802     0.8640  0.86600    0.8640   0.86600   
3                    1106     0.7990  0.80600    0.8000   0.80600   
4                    3148     0.8580  0.86300    0.8590   0.86300   
5                     950     0.9080  0.91600    0.9110   0.91600   
6                     759     0.7220  0.74200    0.7240   0.74200   
7                     572     0.6990  0.71200    0.7030   0.71200   
8                     974     0.6390  0.64600    0.6400   0.64600   
9                    2859     0.9660  0.96900    0.9670   0.96900   
10                   1012     0.6710  0.67900    0.6730   0.67900   
11                    881     0.7680  0.78300    0.7740   0.78300   
12                    716     0.5850  0.60100    0.5850   0.60100   
13                    782     0.6580  0.67300    0.6600   0.67300   
14                   1368     0.5730  0.58400    0.5750   0.58400   
15                   2460     0.8110  0.81700    0.8120   0.81700   
16                    747     0.5970  0.61700    0.6030   0.61700   
17                    840     0.9470  0.96200    0.9550   0.96200   
18                    977     0.9350  0.94900    0.9410   0.94900   
19                   8173     0.9980  0.99900    0.9990   0.99900   
20                      0        NaN      NaN       NaN       NaN   
21                      0        NaN      NaN       NaN       NaN   
22                      1        NaN      NaN       NaN       NaN   
23                  34536     0.7952  0.80485    0.7977   0.80485   

    baseline (majority)  improvement  
0               0.57300       0.9430  
1               0.51200       0.8710  
2               0.49100       0.7370  
3               0.43700       0.6550  
4               0.63500       0.6250  
5               0.77700       0.6230  
6               0.43900       0.5400  
7               0.37800       0.5370  
8               0.26600       0.5170  
9               0.93800       0.4940  
10              0.40400       0.4610  
11              0.60200       0.4560  
12              0.31100       0.4200  
13              0.44500       0.4100  
14              0.30700       0.4000  
15              0.70800       0.3740  
16              0.42300       0.3360  
17              0.94400       0.3190  
18              0.93800       0.1800  
19              0.99900      -0.0000  
20           

In [14]:
result_dt4 = classifier_utils.ProjectsResults(model_3, projects, non_features_columns, replace_na=True)
report_dt4 = result_dt4.get_report_df(include_overall=True)

In [15]:
report_dt4

project  observations  \
0                         CCI-MIT__XCoLab          5512   
1                apache__directory-server           845   
2                        jgralab__jgralab          2072   
3   CloudStack-extras__CloudStack-archive          1424   
4                        Unidata__thredds          1154   
5                        apache__accumulo          4113   
6                         getrailo__railo           815   
7                  Ramblurr__Anki-Android           892   
8                  TeamDev-Ltd__OpenFaces          2979   
9                     apache__lucene-solr          1256   
10                              zkoss__zk          1087   
11                          freenet__fred          1268   
12               hibernate__hibernate-orm          1000   
13                        atlasapi__atlas          1107   
14                 elastic__elasticsearch          1246   
15                           alexo__wro4j          1663   
16                 eclipse__jetty.project          1537   
17      android__platform_frameworks_base          3557   
18                 eucalyptus__eucalyptus          1020   
19                   SINTEF-9012__ThingML          1022   
20                  alkacon__opencms-core           923   
21            revolsys__com.revolsys.open           991   
22                  sebastianbenz__Jnario          8298   
23                                Overall         45781   

    observations (wt NaN)  precision    recall  f1-score  accuracy  \
0                    5512   0.973000  0.977000  0.974000  0.977000   
1                     845   0.929000  0.931000  0.930000  0.931000   
2                    2072   0.865000  0.868000  0.865000  0.868000   
3                    1424   0.816000  0.822000  0.816000  0.822000   
4                    1154   0.907000  0.913000  0.908000  0.913000   
5                    4113   0.871000  0.879000  0.874000  0.879000   
6                     815   0.762000  0.779000  0.769000  0.779000   
7                     892   0.741000  0.763000  0.744000  0.763000   
8                    2979   0.959000  0.962000  0.960000  0.962000   
9                    1256   0.635000  0.640000  0.632000  0.640000   
10                   1087   0.790000  0.808000  0.796000  0.808000   
11                   1268   0.672000  0.681000  0.673000  0.681000   
12                   1000   0.601000  0.613000  0.600000  0.613000   
13                   1107   0.669000  0.688000  0.671000  0.688000   
14                   1246   0.615000  0.608000  0.609000  0.608000   
15                   1663   0.592000  0.601000  0.593000  0.601000   
16                   1537   0.641000  0.658000  0.635000  0.658000   
17                   3557   0.825000  0.834000  0.826000  0.834000   
18                   1020   0.628000  0.651000  0.635000  0.651000   
19                   1022   0.935000  0.945000  0.940000  0.945000   
20                    923   0.927000  0.945000  0.935000  0.945000   
21                    991   0.807000  0.805000  0.745000  0.805000   
22                   8298   0.998000  0.999000  0.999000  0.999000   
23                  45781   0.789478  0.798696  0.788217  0.798696   

    baseline (majority)  improvement  
0              0.454000     0.957000  
1              0.515000     0.859000  
2              0.480000     0.746000  
3              0.453000     0.674000  
4              0.775000     0.615000  
5              0.687000     0.613000  
6              0.459000     0.592000  
7              0.436000     0.581000  
8              0.924000     0.502000  
9              0.289000     0.494000  
10             0.622000     0.491000  
11             0.424000     0.446000  
12             0.312000     0.438000  
13             0.455000     0.428000  
14             0.319000     0.425000  
15             0.310000     0.422000  
16             0.446000     0.383000  
17             0.743000     0.353000  
18             0.463000     0.350000  
19             0.923000    

In [16]:
df_inner_rf = pd.merge(report_dt3, report_dt4, on='project', how='inner', suffixes=('_withoutNA', '_withNA'))

accuracy_rf = df_inner_rf.filter(regex=("project|accuracy.*")).copy()
accuracy_rf['improvement'] = accuracy_rf.apply(lambda x: classifier_utils.get_normalized_improvement(x['accuracy_withNA'], x['accuracy_withoutNA']), axis=1)
accuracy_rf

project  accuracy_withoutNA  \
0                         CCI-MIT__XCoLab             0.97600   
1                apache__directory-server             0.93700   
2                        jgralab__jgralab             0.86600   
3   CloudStack-extras__CloudStack-archive             0.80600   
4                        apache__accumulo             0.86300   
5                        Unidata__thredds             0.91600   
6                  Ramblurr__Anki-Android             0.74200   
7                         getrailo__railo             0.71200   
8                     apache__lucene-solr             0.64600   
9                  TeamDev-Ltd__OpenFaces             0.96900   
10                          freenet__fred             0.67900   
11                              zkoss__zk             0.78300   
12               hibernate__hibernate-orm             0.60100   
13                        atlasapi__atlas             0.67300   
14                           alexo__wro4j             0.58400   
15      android__platform_frameworks_base             0.81700   
16                 eucalyptus__eucalyptus             0.61700   
17                  alkacon__opencms-core             0.96200   
18                   SINTEF-9012__ThingML             0.94900   
19                  sebastianbenz__Jnario             0.99900   
20                 elastic__elasticsearch                 NaN   
21                 eclipse__jetty.project                 NaN   
22            revolsys__com.revolsys.open                 NaN   
23                                Overall             0.80485   

    accuracy_withNA  improvement  
0          0.977000     0.041667  
1          0.931000    -0.006403  
2          0.868000     0.014925  
3          0.822000     0.082474  
4          0.879000     0.116788  
5          0.913000    -0.003275  
6          0.763000     0.081395  
7          0.779000     0.232639  
8          0.640000    -0.009288  
9          0.962000    -0.007224  
10         0.681000     0.006231  
11         0.808000     0.115207  
12         0.613000     0.030075  
13         0.688000     0.045872  
14         0.601000     0.040865  
15         0.834000     0.092896  
16         0.651000     0.088773  
17         0.945000    -0.017672  
18         0.945000    -0.004215  
19         0.999000     0.000000  
20         0.608000          NaN  
21         0.658000          NaN  
22         0.805000          NaN  
23         0.798696    -0.007647

In [21]:
increased_n = len(accuracy_rf[(accuracy_rf['improvement'] > 0) &  (accuracy_rf['project'] != 'Overall')])
decreased_n = len(accuracy_rf[(accuracy_rf['improvement'] < 0) &  (accuracy_rf['project'] != 'Overall')])
constant_n = len(accuracy_rf[(accuracy_rf['improvement'] == 0) &  (accuracy_rf['project'] != 'Overall')])
print(f'Increased in {increased_n} and decreased in {decreased_n} projects. Constant in {constant_n}')

Increased in 13 and decreased in 6 projects. Constant in 1


The overall average accuracy with random forest decreased from 0.80485 to 0.798696. It decreased in 6 of 20 projects, increased in 13 and remained constant in 1.
On the other hand, inputing a constant to replace the NA values allowed to go from 20 to 23 projects.